cpp/src/io/csv/csv_reader_impl.hpp

```diff
@@ -0,0 +1,132 @@
+  std::vector<char> header;           ///< Header row data, for parsing column names.
+
+  // Specifying which part of the file to parse
+  size_t byte_range_offset = 0; ///< Offset into the data to start parsing.
+  size_t byte_range_size = 0;   ///< Length of the data of interest to parse.
+  gdf_size_type nrows = -1;     ///< Number of rows to read. -1 for all rows.
+  gdf_size_type skiprows = 0;   ///< Number of rows to skip from the start.
+  gdf_size_type skipfooter = 0; ///< Number of rows to skip from the end.
+
+
+  void setColumnNamesFromCsv();
+  void countRecordsAndQuotes(const char *h_data, size_t h_size);
+  void setRecordStarts(const char *h_data, size_t h_size);
+  void uploadDataToDevice(const char *h_uncomp_data, size_t h_uncomp_size);
+
+  void launch_dataConvertColumns(void **d_gdf, gdf_valid_type **valid, gdf_dtype *d_dtypes, gdf_size_type *num_valid);
+  void launch_dataTypeDetection(column_data_t *d_columnData);
+
+public:
+  /**---------------------------------------------------------------------------*
+   * @brief CsvReader::Impl constructor; throws if arguments are not supported
+   *---------------------------------------------------------------------------**/
+  explicit Impl(csv_reader_args const &args);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Parse the input CSV file as specified with the args_ data member
+   *
+   * @return cudf::table object that contains the array of gdf_columns
+   *---------------------------------------------------------------------------**/
+  table read();
+
+  /**---------------------------------------------------------------------------*
+   * @brief Parse the specified byte range of the input CSV file.
+   *
+   * Reads the row that starts before or at the end of the range, even if it ends
+   * after the end of the range.
+   *
+   * @param[in] offset Offset of the byte range to read.
+   * @param[in] size Size of the byte range to read. Set to zero to read to
+   * the end of the file.
+   *
+   * @return cudf::table object that contains the array of gdf_columns
+   *---------------------------------------------------------------------------**/
+  table read_byte_range(size_t offset, size_t size);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Parse the specified rows of the input CSV file.
+   * 
+   * Set num_skip_footer to zero when using num_read parameter.
+   *
+   * @param[in] num_skip_header Number of rows at the start of the files to skip.
+   * @param[in] num_skip_footer Number of rows at the bottom of the file to skip.
+   * @param[in] num_read Number of rows to read. Value of -1 indicates all rows.
+   * 
+   * @return cudf::table object that contains the array of gdf_columns
+   *---------------------------------------------------------------------------**/
+  table read_rows(gdf_size_type num_skip_header, gdf_size_type num_skip_footer, gdf_size_type num_read);
```

Comment:
```suggestion
  table read_rows(gdf_size_type num_skip_header, gdf_size_type num_skip_footer, gdf_size_type num_rows);
```

Prefer using `num_rows` for commonality with Parquet and ORC.